core/src/main/java/de/danoeh/antennapod/core/sync/EpisodeActionFilter.java

```diff
@@ -0,0 +1,88 @@
+package de.danoeh.antennapod.core.sync;
+
+import android.util.Log;
+
+import androidx.collection.ArrayMap;
+import androidx.core.util.Pair;
+
+import java.util.List;
+import java.util.Map;
+
+import de.danoeh.antennapod.net.sync.model.EpisodeAction;
+
+public class EpisodeActionFilter {
+
+    public static final String TAG = "EpisodeActionFilter";
+
+    public static Map<Pair<String, String>, EpisodeAction> getRemoteActionsOverridingLocalActions(
+            List<EpisodeAction> remoteActions,
+            List<EpisodeAction> queuedEpisodeActions) {
+        // make sure more recent local actions are not overwritten by older remote actions
+        Map<Pair<String, String>, EpisodeAction> remoteActionsThatOverrideLocalActions = new ArrayMap<>();
+        Map<Pair<String, String>, EpisodeAction> localMostRecentPlayActions =
+                createUniqueLocalMostRecentPlayActions(queuedEpisodeActions);
+        for (EpisodeAction remoteAction : remoteActions) {
+            Log.d(TAG, "Processing remoteAction: " + remoteAction.toString());
+            Pair<String, String> key = new Pair<>(remoteAction.getPodcast(), remoteAction.getEpisode());
+            switch (remoteAction.getAction()) {
+                case NEW:
+                    remoteActionsThatOverrideLocalActions.put(key, remoteAction);
+                    break;
+                case DOWNLOAD:
+                    break;
+                case PLAY:
+                    EpisodeAction localMostRecent = localMostRecentPlayActions.get(key);
+                    if (localActionHappenedAfterRemoteAction(remoteAction, localMostRecent)) {
+                        break;
+                    }
+                    EpisodeAction mostRecentAction = remoteActionsThatOverrideLocalActions.get(key);
+                    if (mostRecentAction == null
```

Comment:
Don't you check for null in the method already?

Reply:
in what method?